cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputFormat.java

```diff
@@ -0,0 +1,201 @@
+
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+
+import static org.apache.hadoop.mapreduce.TaskType.MAP;
+
+/**
+ * This class extends the FileOutputFormat and allows writing dynamically to multiple partitions of a PartitionedFileSet
+ * Dataset.
+ *
+ * @param <K> Type of key
+ * @param <V> Type of value
+ */
+public class DynamicPartitioningOutputFormat<K, V> extends FileOutputFormat<K, V> {
+
+  private FileOutputFormat<K, V> fileOutputFormat;
+
+  /**
+   * Create a composite record writer that can write key/value data to different output files.
+   *
+   * @return a composite record writer
+   * @throws IOException
+   */
+  @Override
+  public RecordWriter<K, V> getRecordWriter(final TaskAttemptContext job) throws IOException {
+    final String outputName = FileOutputFormat.getOutputName(job);
+
+    Class<? extends DynamicPartitioner> className = job.getConfiguration()
+      .getClass(PartitionedFileSetArguments.DYNAMIC_PARTITIONER_CLASS_NAME, null, DynamicPartitioner.class);
+
+    @SuppressWarnings("unchecked")
+    final DynamicPartitioner<K, V> dynamicPartitioner =
+      new InstantiatorFactory(false).get(TypeToken.of(className)).create();
+
+    MapReduceMetrics.TaskType taskType = MapReduceMetrics.TaskType.from(job.getTaskAttemptID().getTaskType());
+    final BasicMapReduceTaskContext<K, V> mrTaskContext = new MapReduceTaskContextProvider(job, taskType).get();
+    String outputDatasetName = job.getConfiguration().get(Constants.Dataset.Partitioned.HCONF_ATTR_OUTPUT_DATASET);
+    PartitionedFileSet outputDataset = mrTaskContext.getDataset(outputDatasetName);
+    final Partitioning partitioning = outputDataset.getPartitioning();
+
+    dynamicPartitioner.initialize(mrTaskContext);
+
+    return new RecordWriter<K, V>() {
+
+      // a cache storing the record writers for different output files.
+      Map<PartitionKey, RecordWriter<K, V>> recordWriters = new HashMap<>();
+
+      public void write(K key, V value) throws IOException, InterruptedException {
+        PartitionKey partitionKey = dynamicPartitioner.getPartitionKey(key, value);
+        RecordWriter<K, V> rw = this.recordWriters.get(partitionKey);
+        if (rw == null) {
+          String relativePath = PartitionedFileSetDataset.getOutputPath(partitionKey, partitioning);
+          String finalPath = relativePath + "/" + outputName;
+
+          // if we don't have the record writer yet for the final path, create one and add it to the cache
+          rw = getBaseRecordWriter(getTaskAttemptContext(job, finalPath));
+          this.recordWriters.put(partitionKey, rw);
+        }
+        rw.write(key, value);
+      }
+
+      @Override
+      public void close(TaskAttemptContext context) throws IOException, InterruptedException {
+        try {
+          for (RecordWriter<K, V> recordWriter : this.recordWriters.values()) {
+            recordWriter.close(context);
```

Comment:
I put a static method in MultipleOutputs to close all RecordWriters, and utilized it here now.